cognite/client/data_classes/datapoints.py

```diff
@@ -271,19 +372,66 @@ def __str__(self):
+    ) -> "pandas.DataFrame":
         """Convert the datapoints list into a pandas DataFrame.
 
         Args:
             column_names (str): Which field to use as column header. Defaults to "externalId", can also be "id".
+            include_aggregate_name (bool): Include aggregate in the column name
+            complete (str): Post-processing of the dataframe.
+
+                Pass 'fill' to insert missing entries into the index, and complete data where possible (supports interpolation, stepInterpolation, count, sum, totalVariation).
+
+                Pass 'fill,dropna' to additionally drop rows in which any aggregate for any time series has missing values (typically rows at the start and end for interpolation aggregates).
+                This option guarantees that all returned dataframes have the exact same shape and no missing values anywhere, and is only supported for aggregates sum, count, totalVariance, interpolation and stepInterpolation.
+
         Returns:
             pandas.DataFrame: The datapoints list as a pandas DataFrame.
         """
         pd = utils._auxiliary.local_import("pandas")
-        dfs = [df.to_pandas(column_names=column_names) for df in self.data]
-        if dfs:
-            return pd.concat(dfs, axis="columns")
-        return pd.DataFrame()
+
+        complete = Datapoints._get_complete_options(complete)
+        if set(complete) - set(["fill", "dropna", ""]):
+            raise ValueError("complete should be 'fill', 'fill,dropna' or Falsy")
+        fill_flag = "fill" in complete
+
+        dfs = self._filled_dataframe(column_names, include_aggregate_name, fill_flag)
+        if not dfs:
+            return pd.DataFrame()
+
+        df = pd.concat(dfs, axis="columns")
+        if "dropna" in complete:
+            df = Datapoints._dropna_dataframe(
+                df, [ag for df in self.data for ag, _ in df._get_non_empty_data_fields(get_empty_lists=True)]
+            )
+
+        return df
+
+    def _filled_dataframe(self, column_names, include_aggregate_name, fill):
+        is_step_dict = {dp.id: False for dp in self.data}
+        complete_start = None
+        complete_end = None
+        if fill:
+            interp_ids = [dp.id for dp in self.data if dp.interpolation is not None]
+            starts = [dp.timestamp[0] for dp in self.data if dp.timestamp]
+            ends = [dp.timestamp[-1] for dp in self.data if dp.timestamp]
+            complete_start = min(starts) if starts else None
+            complete_end = max(ends) if starts else None
+            if interp_ids:  # get isStep flag and pass it to completion to handle interpolation correctly
+                is_step_dict.update(Datapoints._is_step(self._cognite_client, interp_ids))
+        return [
+            df.to_pandas(
+                column_names=column_names,
+                include_aggregate_name=include_aggregate_name,
+                complete="fill" if fill else None,  # don't pass dropna
```

Comment:
Wont `fill` here still always evaluate to `True` since you are doing `if fill` on line 416?

Reply:
this is not in the block starting at 416